### PR TITLE
feat(plugin-nm): support configuring link mode

### PR DESCRIPTION
**Pull request by jeff-wishnie**
**What's the problem this PR addresses?**

Closes yarnpkg/berry#4979

Currently on Windows `yarn` uses `junctions` to link workspaces into `node_modules` directories. On other platforms `yarn` uses symlinks.

As a result, workspace links in `node_modules` on Windows are _always_ absolute paths. On other platforms links may be relative paths, leading to _inconsistent behavior_ when using `nodeLinker: node-modules` and running `yarn install` from Windows vs. Linux.

A particularly bad side-effect of absolute paths on Windows is that it breaks Docker builds--when a monorepo is mounted into a container, the absolute paths are invalid.

The PR adds an additional configuration option for the `nm-linker` to use symbolic links on Windows as well, thereby supporting relative links, and fixing container builds.
...

**How did you fix it?**
added the option `nmFolderLinkMode` with two settiings:
- `classic` (the default) -- `nm-linker` behaves exactly as today, using junctions on Windows and symlinks on other platforms.
- `symlinks` -- `nm-linker` uses symlinks on _all_ platforms

...

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective. 
       **[PLEASE CHECK MY WORK HERE!]**

- [x] I will check that all automated PR checks pass before the PR gets reviewed. 
       **[A handful of unit and integration tests are failing when I run against `master`. This same set fails on my branch]**

### CODE REVIEW

**Comment by jeff-wishnie** [comment]
There seems to be an issue with the `E2E NM Babel Install` pipeline. There hasn't been a passing run since Sept 18.

**Comment by merceyz** [comment]
Indeed, they have some custom plugins that aren't compatible with Yarn v4 (`master`).

File: packages/plugin-nm/sources/NodeModulesLinker.ts
Comment (merceyz) [inline packages/plugin-nm/sources/NodeModulesLinker.ts:682]:
```suggestion
    if (!stats || stats.isDirectory()) {
```
The "parent" if statement already checked that it's `win32`

**Review by merceyz — DISMISSED** [review]

**Review by merceyz — COMMENTED** [review]
We also do this in the pnpm linker, should the same logic be applied there?
https://github.com/yarnpkg/berry/blob/63a26fec156d9f4cc3e924dbe8ba2e6cd67899b0/packages/plugin-pnpm/sources/PnpmLinker.ts#L235

File: packages/plugin-nm/sources/NodeModulesLinker.ts
Comment (jeff-wishnie) [reply packages/plugin-nm/sources/NodeModulesLinker.ts:682]: Thank you! Fixed!

**Comment by jeff-wishnie** [comment]
> We also do this in the pnpm linker, should the same logic be applied there?
> 
> https://github.com/yarnpkg/berry/blob/63a26fec156d9f4cc3e924dbe8ba2e6cd67899b0/packages/plugin-pnpm/sources/PnpmLinker.ts#L235

That probably would make sense, but I'm not familiar with `pnpm` (never used it) so wouldn't want to make that call.

If the `pnpm` folks like it's the right thing to do, would it make sense to have a single configuration setting that applies to both `nm-linker` and `pnpm-linker` or separate flags? It seems like the convention is for plugins to have their own config extensions (e.g. the configs prefixes with ` nm` or `npm`) and that it could get messy if plugins tried to share config settings that are not defined in core.

It looks like you are one of the maintainers of `pnpm-plugin`, what would you like to do? 

I'm happy to make a similar PR for that plugin.

File: packages/plugin-nm/sources/NodeModulesLinker.ts
Comment (larixer) [inline packages/plugin-nm/sources/NodeModulesLinker.ts:1283]: There is no need to make `nmFolderLinkMode` an object with this structure. Don't look at `nmMode` since, in case of `nmMode` it might be dynamically changed during code execution when certain factors become true. In case of `nmFolderLinkMode` - it never changes, so need to wrap it into an object.

**Comment by larixer** [comment]
The integration test for this feature is missing, we typically have integration tests for all the features in Yarn, that make sure they will not become broken in the future. Please add a test.

File: packages/plugin-nm/sources/NodeModulesLinker.ts
Comment (jeff-wishnie) [reply packages/plugin-nm/sources/NodeModulesLinker.ts:1283]: Thanks, was wondering why nmMode used this structure and just copied the patterns I saw in the code. 

I'll update the PR.

File: packages/acceptance-tests/pkg-tests-specs/sources/node-modules.test.ts
Comment (jeff-wishnie) [inline packages/acceptance-tests/pkg-tests-specs/sources/node-modules.test.ts:5]: Is this already present in a helper lib somewhere?

File: packages/acceptance-tests/pkg-tests-specs/sources/node-modules.test.ts
Comment (jeff-wishnie) [inline packages/acceptance-tests/pkg-tests-specs/sources/node-modules.test.ts:1852]: Couldn't find a cleaner way to distinguish between a `junction` and `symlink` on Windows. Have any ideas?

Comment (larixer) [reply packages/acceptance-tests/pkg-tests-specs/sources/node-modules.test.ts:1852]: I think there is no cleaner way, as per:
https://github.com/nodejs/help/issues/1520

File: packages/acceptance-tests/pkg-tests-specs/sources/node-modules.test.ts
Comment (merceyz) [reply packages/acceptance-tests/pkg-tests-specs/sources/node-modules.test.ts:5]: We have a portable shell (`@yarnpkg/shell`) but since this is for a Windows specific test it's fine to use `exec` like this.

**Comment by jeff-wishnie** [comment]
@larixer @merceyz , thank you for the reviews.

@merceyz I think merge is blocked by the change you requested that I made via a push rather than the GitHub UI. I don't see how to clear the change-request, is that something you can do?

**Comment by jeff-wishnie** [comment]
I think this is ready for merge.

@merceyz would you like me to propose a similar PR for the pnpm linker?

**Review by merceyz — COMMENTED** [review]
> [...], what would you like to do?
> would you like me to propose a similar PR for the pnpm linker?

I was thinking there should be a single setting controlling all symlink/junction creation in Yarn instead of one for each individual case however I wont block this PR on that.

**Comment by jeff-wishnie** [comment]
> I was thinking there should be a single setting controlling all symlink/junction creation in Yarn

I'd be happy to work on that as a follow-up PR:
- replace current nm-linker specific config with a global setting
- update the pnpm linker to respect that setting as well.

I don't see junctions used elsewhere in the code base, but wonder if there are any contrib plugins that do that would want to update as well?

**Comment by larixer** [comment]
@jeff-wishnie Sounds good. I'm not aware about any contrib plugins. I think the `pnpm` linker is the only other place where `junctions` are used. We would be grateful if you submit a follow-up PR for `pnpm` linker.

**Comment by jeff-wishnie** [comment]
Will! As noted above, I've never worked with `pnpm`, this seems a pretty mechanical change so I'm not concerned but definitely looking for another solid review of that one!

**Pull request merged**
